tv_apps/tv-deck/js/channel_manager.js

```diff
@@ -0,0 +1,158 @@
+    this.pinnedChannels = {};
+    this.cardManager.getCardList().then(function(cardList) {
+      cardList.forEach(function(card) {
+        var url = card.launchURL;
+        if (url && url.indexOf('app://tv-deck.gaiamobile.org') === 0) {
+          var channelHash = url.substring(url.indexOf('#') + 1);
+          this.pinnedChannels[channelHash] = true;
+          return true;
+        }
+        return false;
+      }.bind(this));
+      this.updatePinButton();
+    }.bind(this));
+  };
+
+  /**
+   * Update pin button in button-group-panel and contextmenu.
+   */
+  proto.updatePinButton = function cm_updatePinButton() {
+    if (this.pinnedChannels[this._tvDeck.currentHash]) {
+      // Show unpin button if current channel is pinned.
+      this._tvDeck.pinButtonContextmenu.setAttribute(
+                                           'data-l10n-id', 'unpin-from-home');
+      this._tvDeck.pinButtonContextmenu.onclick =
+                                           this.unpinFromHome.bind(this);
+      this._tvDeck.pinButton.onclick = this.unpinFromHome.bind(this);
+    } else {
+      // Show pin button if current channel is not pinned yet.
+      this._tvDeck.pinButtonContextmenu.setAttribute(
+                                           'data-l10n-id', 'pin-to-home');
+      this._tvDeck.pinButtonContextmenu.onclick = this.pinToHome.bind(this);
+      this._tvDeck.pinButton.onclick = this.pinToHome.bind(this);
+    }
+  };
+
+  proto.pinToHome = function cm_pinToHome() {
+    if (!this._tvDeck.playingTunerId) {
+      return;
+    }
+
+    var title = this._tvDeck.playingChannel.channel.name;
+    var number = this._tvDeck.playingChannelNumber;
+
+    this.createChannelBlob(title, number, function(blob) {
+      /* jshint nonew:false */
+      new MozActivity({
+        name: 'pin',
+        data: {
+          type: 'Application',
+          group: 'tv',
+          name: number,
+          manifestURL: this._tvDeck.manifestURL,
+          launchURL: window.location.href,
+          thumbnail: blob
+        }
+      });
+    }.bind(this));
+  };
+
+  proto.unpinFromHome = function cm_unpinFromHome() {
+
+    var message = {
+      type: 'unpin',
+      data: {
+        manifestURL: this._tvDeck.manifestURL,
+        launchURL: window.location.href
+      }
+    };
+
+    navigator.mozApps.getSelf().onsuccess = function(evt) {
+      var selfApp = evt.target.result;
+      selfApp.connect('appdeck-channel').then(function (ports) {
+        ports.forEach(function(port) {
+          port.postMessage(message);
+        });
+      });
+    }.bind(this);
+  };
+
+  proto.createChannelBlob =
+                       function tc_createChannelBlob(title, number, callback) {
+    var canvas = document.createElement('canvas');
+    canvas.width = 334;
+    canvas.height = 334;
+    var ctx = canvas.getContext('2d');
+
+    var data = '<svg xmlns="http://www.w3.org/2000/svg" width="334" ' +
```

Comment:
It looks like that you draw a circle with title and CH + number. May you use canvas to draw everything instead of load SVG into image and draw it into canvas?